### ISSUE TITLE
Test We Vote with a Cross-browser tool

### ISSUE THREAD

**Issue by DaleMcGrew** [wevote/WebApp#515]
We should run some tests of our HTML/CSS with one of these cross-browser tools:
http://mashable.com/2014/02/26/browser-testing-tools

**Comment by luojxxx** [comment]
I put together a little guide for using cross browser testing cloud services with Selenium's automated testing: https://docs.google.com/document/d/1gDi1eXimhN-7syTt-OK9V6elBWC8F9s9riz_0qH6nek/edit?usp=sharing

It's fairly straightforward it seems. You can use the Selenium Firefox extension to record the website test actions, convert them into Selenium WebDriver format, modify that code to use the Saucelabs/someOtherService cloud, and then watch it on the Saucelabs/someOtherService.

**Comment by ixley** [comment]
Related: Establish list of supported browsers: https://github.com/wevote/WebApp/issues/522

**Comment by pertrai1** [comment]
Selenium testing works great but the drivers are limited for cross browser testing. The simplest way would be to use the device toolbar if working in Chrome or Responsive Design Mode in Firefox. Continually change the different devices to get a close idea of simulation on how the site will look being that it is a responsive app.

**Comment by flobotbear** [comment]
Not sure if this issue is still relevant, but Browserstack does offer its services for free of charge to OSS projects https://www.browserstack.com/pricing

**Comment by alexthornton** [comment]
Cross-browser testing is a good idea, but I think the priorities might be a little misplaced. There don't appear to be any tests written yet (https://github.com/wevote/WebApp/tree/master/tests). In a hierarchy of testing, I think we would want to worry about cross-browser testing after we have basic unit and functional tests for the application. If the application logic breaks, it's broken in every browser, so testing in React first will probably yield more bang for our buck compared to cross-browser testing.

**Comment by Anoninnyc** [comment]
For now.... they offer a free trial feature on their homepage.

**Comment by ixley** [comment]
I agree with @alexthornton's point about priority, and that we should put more focus on automated testing of app functionality at this stage.

I *do* think that it will be helpful to have a clear sense of which browsers we officially support, but I don't see such a high value for browser testing at this stage because it's not clear how we would automate such a process and unless there are dedicated volunteers and an established QA process, it seems unlikely to get checked with any regularity. Because the project is currently geared towards embracing newer technologies and supporting mostly contemporary browsers, there is also lower likelihood of regular browser testing identifying breakage on individual browsers. *And without unit tests or visual regression tests, automation is much more difficult.*

All this is not to say there is NO value, just that it might be disproportionate to the (mostly manual) effort required. If there are folks that do want to move forward with this effort, I would encourage creating a QA process document that is clear and easy to follow and cite cases for when and how to browser test new features, releases, the app in general, etc.

**Comment by pertrai1** [comment]
There might be some confusion in regards to what these different levels of testing are. Cross browser testing is for the most part manual testing. Functional testing is where there are selenium test scripts that are created and run in the CI. @josephevans and I started the process so there would be a grid that would run this in the cloud during the CI process and depending on the tests would depend on the progress of the process. Most of the time these functional tests are written based on the issue being handled here or in whatever agile board you take care of stories and tasks.

Unit testing at this point is a little more trickier. I can agree with the fact that unit tests can be written for any code that is developed from this point forward, but not sure about code that has already been developed. Those tests would be more conformation tests than unit tests if you go by the red,green,red,green approach. Don't let me discourage anyone from unit testing, as the definition of "done" in agile requires unit testing. 

For my two cents - or one depending on my worth - is that a setup is focused on a good suite of functional tests that start to run in CI based off of using selenium. The trick here is which domain specific language you write the tests in. In the past I have used Ruby to write the tests, but have also done some smaller ones in Java. I do know there are some, such as the end to end suite that we have using Angular. 

Anyways, that is a direction I would consider as you move forward with building a stronger foundation in the CI process.
